### PR TITLE
fix: MISRA C:2012 Rule 10.3 compliance - narrowing conversions (Issue #845)

### PR DESCRIPTION
**Pull request by jlaustill**
## Summary
- Fixes all ~106 MISRA 10.3 violations for narrowing conversions and cross-type-category assignments
- Adds explicit casts where C's implicit type conversions would violate MISRA compliance
- Handles edge cases like `-2147483648` literal overflow and float/int cross-category conversions

## Test plan
- [x] All 941 integration tests pass
- [x] All 5452 unit tests pass
- [x] MISRA 10.3 violations: 106 → 0 (100% fixed)
- [x] `npm run validate:c` confirms 0 Rule 10.3 violations

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/transpiler/output/codegen/assignment/handlers/SimpleHandler.ts [hunk 1]
```diff
@@ -3,25 +3,92 @@
...
+import TYPE_MAP from "../../types/TYPE_MAP.js";
+import CppModeHelper from "../../helpers/CppModeHelper.js";
 
 /** Get typed generator reference */
 function gen(): ICodeGenApi {
   return CodeGenState.generator as ICodeGenApi;
 }
 
+/**
+ * Map compound assignment operator to its binary operator equivalent.
```
Comment (jlaustill) [inline src/transpiler/output/codegen/assignment/handlers/SimpleHandler.ts]: **[important]** design: This `COMPOUND_TO_BINARY` map duplicates operator knowledge. Check if a similar mapping already exists in `AssignmentHandlerUtils.ts` or elsewhere in the codegen — per project guidelines, having to update the same information in multiple places is an anti-pattern.

If no existing mapping, consider extracting this to a shared constant in the `types/` or `constants/` directory so it can be reused.

File: src/utils/BitUtils.ts [hunk 2]
```diff
@@ -102,6 +111,22 @@ class BitUtils {
...
 
+  /**
+   * Wrap an expression with a cast for MISRA 10.3 compliance on narrow types.
+   * Bit manipulation operations promote to int; this casts back to the target type.
+   *
+   * @param expr - The expression to potentially wrap
+   * @param targetType - The C-Next type name (e.g., "u8", "u16")
+   * @returns Expression wrapped with cast if narrow, or original expression
+   */
+  private static wrapNarrowCast(expr: string, targetType?: string): string {
```
Comment (jlaustill) [inline src/utils/BitUtils.ts]: **[moderate]** design: This `wrapNarrowCast` method duplicates logic from `NarrowingCastHelper.wrap`. Per project guidelines, dual code paths are the "WORST anti-pattern." Consider:

1. Using `NarrowingCastHelper.wrap` here instead, or
2. If there's a reason for separate implementations (e.g., avoiding circular deps), document why in a comment

Currently both check `NARROW_TYPES` and apply the same cast pattern, so updating one requires updating both.

File: src/transpiler/output/codegen/generators/expressions/UnaryExprGenerator.ts [hunk 3]
```diff
@@ -45,7 +54,22 @@ const generateUnaryExpr = (
...
-  if (text.startsWith("-")) return { code: `-${inner}`, effects: [] };
+  if (text.startsWith("-")) {
+    // MISRA 10.3: Handle problematic negative literals that overflow in C
+    // -2147483648 is parsed as -(2147483648) where 2147483648 > INT32_MAX
+    // Must use INT32_MIN or INT64_MIN to avoid the overflow
+    // Cast is needed because INT32_MIN has type 'int', not 'int32_t'
+    const effects: TGeneratorEffect[] = [];
+    if (inner === INT32_MIN_LITERAL) {
+      effects.push({ type: "include", header: "limits" });
+      return { code: "(int32_t)INT32_MIN", effects };
```
Comment (jlaustill) [inline src/transpiler/output/codegen/generators/expressions/UnaryExprGenerator.ts]: Nit: The string concatenation `INT64_MIN_LITERAL + "LL"` is evaluated at runtime. Consider using a const:

```suggestion
const INT64_MIN_LITERAL_SUFFIXED = "9223372036854775808LL";
```

Also: is it possible to receive `9223372036854775808LL` as `inner` here? If the lexer strips suffixes, this branch may be unreachable.

**Review by jlaustill — COMMENTED** [review]
## Review Summary

**PR:** #946 — fix: MISRA C:2012 Rule 10.3 compliance - narrowing conversions (Issue #845)
**Reviewed by:** Claude (automated review)

### Overview
This PR adds MISRA C:2012 Rule 10.3 compliance by introducing explicit casts when assigning results of bit operations (which C promotes to `int`) back to narrower types. The implementation is well-structured with a new `NarrowingCastHelper` utility class and consistent integration across the code generation pipeline.

### Findings

| Severity | Count |
|----------|-------|
| Critical | 0 |
| Important | 2 |
| Moderate | 1 |
| Nit | 2 |

### Key Issues

1. **[important]** The `BinaryExprGenerator.ts` and other expression generators may also need narrowing cast handling for arithmetic operations that produce `int` results assigned to narrow types. Verify these paths don't introduce new MISRA 10.3 violations.

2. **[important]** The `SimpleHandler.ts` COMPOUND_TO_BINARY map is duplicated logic — similar operator mappings likely exist elsewhere (e.g., `AssignmentHandlerUtils.ts` or codegen utilities). Consider extracting to shared constant.

3. **[moderate]** The `BitUtils.wrapNarrowCast` method has similar logic to `NarrowingCastHelper.wrap` — potential dual code path issue per project guidelines.

### What Looks Good

- **Excellent test coverage**: Unit tests for `NarrowingCastHelper`, `BitRangeHelper`, and `ArrayAccessHelper` thoroughly cover the new functionality
- **Clean architecture**: The `NarrowingCastHelper` utility follows the project's static class pattern and is well-documented
- **Comprehensive scope**: Handles bit extraction, character literals, compound assignments, and cross-type-category conversions
- **Good documentation**: Implementation plan in `docs/plans/` explains the reasoning clearly

---
🤖 *Automated review by Claude Code*

**Comment by jlaustill** [comment]
## Review Feedback Response

**Issue 1: BinaryExprGenerator may need narrowing cast handling**
✅ Not needed. MISRA C:2012 Rule 10.3 violations are handled at the assignment site, not at the expression generation site. Binary expressions produce `int` results, and the narrowing cast is applied when that result is:
- Assigned to a variable (handled by `SimpleHandler.ts` for compound assignments)
- Used to initialize a variable (handled by `VariableDeclHelper.ts`)

Verified: 0 MISRA 10.3 violations in generated C code.

**Issue 2: COMPOUND_TO_BINARY map duplication**
✅ Fixed. Extracted to shared `src/transpiler/output/codegen/types/COMPOUND_TO_BINARY.ts` and updated both `SimpleHandler.ts` and `AtomicGenerator.ts` to use the shared constant.

**Issue 3: BitUtils.wrapNarrowCast similar to NarrowingCastHelper.wrap**
✅ Not a duplicate code path. These serve different purposes:
- `BitUtils.wrapNarrowCast`: Private helper for bit operations, only handles NARROW_TYPES (u8, u16, i8, i16)
- `NarrowingCastHelper.wrap`: Public utility with complex type promotion rules, int-to-bool conversion, cross-type-category handling

The layering is intentional: BitUtils handles low-level bit manipulation, NarrowingCastHelper handles MISRA compliance at the assignment/declaration level.

**Additional fix:** Refactored `handleSimpleAssignment` in `SimpleHandler.ts` to reduce cognitive complexity from 17 to under 15 (extracted two helper functions).
